### ISSUE TITLE
Suggestion: Lock radius when changing brushes (with working patch)

### ISSUE THREAD

**Issue by ChengduLittleA** [mypaint/mypaint#1157]
Hi developers and artists...

In my painting I switch between opaque and smudge brushes frequently, so I think it should be best if we have a "lock radius" function that allows quick change of brush without affecting size (because you will be painting repeatedly on large or small surfaces, this way you won't need to press buttons repeatedly to restore a small or big size each time you change brushes.

So I made this option, it says in the brush menu at the moment:

![lock_radius](https://user-images.githubusercontent.com/7886249/133923081-ee2efe70-61ac-4c08-ab23-846baad2ba4d.png)

And here's the patch:

```
diff --git a/gui/brushmodifier.py b/gui/brushmodifier.py
index f2b46e95..6a5a609f 100644
--- a/gui/brushmodifier.py
+++ b/gui/brushmodifier.py
@@ -96,12 +96,20 @@ class BrushModifier (object):
         b = self.app.brush
         prev_lock_alpha = b.is_alpha_locked()
 
+        lock_radius = self.app.preferences['misc.context_lock_radius']
+        if lock_radius:
+            prev_radius = b.get_base_value('radius_logarithmic')
+
         # Changing the effective brush
         b.begin_atomic()
         color = b.get_color_hsv()
 
         mix_old = b.get_base_value('restore_color')
         b.load_from_brushinfo(brushinfo)
+
+        if lock_radius:
+            b.set_base_value('radius_logarithmic', prev_radius)
+
         self.unmodified_brushinfo = b.clone()
 
         # Preserve color
diff --git a/gui/document.py b/gui/document.py
index 047fca9e..ff911ba4 100644
--- a/gui/document.py
+++ b/gui/document.py
@@ -354,6 +354,11 @@ class Document (CanvasController):  # TODO: rename to "DocumentController"
             toggle_action.set_active(
                 self.app.preferences['misc.context_restores_color']
             )
+            
+            toggle_action = self.app.builder.get_object('ContextLockRadius')
+            toggle_action.set_active(
+                self.app.preferences['misc.context_lock_radius']
+            )
 
         #: Saved transformation to allow FitView to be toggled.
         self.saved_view = None
@@ -1686,6 +1691,11 @@ class Document (CanvasController):  # TODO: rename to "DocumentController"
         value = bool(action.get_active())
         self.app.preferences['misc.context_restores_color'] = value
 
+    def context_toggle_lock_radius_cb(self, action):
+        """GtkToggleAction callback for whether the radius is locked between brush changes"""
+        value = bool(action.get_active())
+        self.app.preferences['misc.context_lock_radius'] = value
+
     ## UI feedback for current layer/stroke
 
     def strokeblink_state_enter(self, strokeshape, bbox, center):
diff --git a/gui/menu.xml b/gui/menu.xml
index eab7a7d1..c20d6759 100644
--- a/gui/menu.xml
+++ b/gui/menu.xml
@@ -142,6 +142,7 @@
         <menuitem action='Context09s'/>
       </menu>
       <menuitem action='PickContext'/>
+      <menuitem action='ContextLockRadius'/>
       <!-- Brush pack import -->
       <separator/>
       <menuitem action='ImportBrushPack'/>
diff --git a/gui/resources.xml b/gui/resources.xml
index a0641e5e..1812dc35 100644
--- a/gui/resources.xml
+++ b/gui/resources.xml
@@ -385,6 +385,13 @@ Vocabulary
         </object>
         <accelerator key="grave"/>
       </child>
+      <child>
+        <object class="GtkToggleAction" id="ContextLockRadius">
+          <property name="label" translatable="yes" context="Menu→Brush">Lock Radius</property>
+          <property name="tooltip" translatable="yes" context="Accel Editor (descriptions)">Toggle whether changing the brush should keep current radius.</property>
+          <signal name="activate" handler="context_toggle_lock_radius_cb"/>
+        </object>
+      </child>
       <child>
         <object class="GtkAction" id="ContextStore">
           <property name="label" translatable="yes" context="Menu→Brush→Shortcut Keys (labels), Accel Editor (labels)">Save Brush to Most Recent Shortcut Key</property>
diff --git a/gui/userconfig.py b/gui/userconfig.py
index 582d7f13..9de0c696 100644
--- a/gui/userconfig.py
+++ b/gui/userconfig.py
@@ -91,6 +91,7 @@ def default_configuration():
         'brushmanager.selected_groups': [],
         'frame.color_rgba': (0.12, 0.12, 0.12, 0.92),
         'misc.context_restores_color': True,
+        'misc.context_lock_radius': False,
 
         'document.autosave_backups': True,
         'document.autosave_interval': 10,
```
Thanks! I think this quite useful, krita and blender can do similar stuff like this as well, really makes things convenient :D

**Comment by odysseywestra** [comment]
This might be better if it was in the tool options box so that the user can see it. Plus you will want to hook it into the keymaping so it can be turn off an on with a keystroke. Not a bad idea overall though.

**Comment by ChengduLittleA** [comment]
Yes, I registered this in the action XML file and the option is visible in the key map editor.
I have yet to look into the toolbar ui, it should be pretty straightforward. And then we might also need a new UI as well.

**Comment by odysseywestra** [comment]
Okay I gotten some time to play with it, and I can really see a use case on this with keeping radius constant when you change brushes. One thing I will note is that the user can still adjust the radius via the brush editor, shortcuts, pointer button shortcuts and options toolbox. Not saying that's a bad thing, just something to note.

With that said, yes there definitely needs to be some feedback to the user such as:
  - The feedback function via lib/feedback.py so you could hook into that to let the user know the brush radius is locked. (Honestly this function is really not being utilized a bunch which it really should for other functions like lock alpha and blending eraser, but I'll digress that to another conversation.)
  - If I recall, MyPaint also use software cursors as well so you can change the symbol to represent that the radius is locked as well.
  - Also yes you can also use the toolbar ui as well to provide the feedback as well. I would put it next to the Eraser function. (I do agree that the toolbar UI needs a rework to make it much more customizable or just allow more functions to be added.)
  - Finally you should add it to the Tool Options for Freehand Drawing, Lines and Curves, Connected Lines, and Circles. To start I'd say a check mark option next to the radius or under would work, but you will have have to play with that.

There's also the argument of applying this to other brush parameters as well. In the toolbox options, there are eight common ones you could do this too for freehand drawing as well. Not to make your patch any more complicated, but it is something to think about.

Also if you not so shy we do have a discord server over at https://discord.gg/vbB434p if you want to continue the conversation there as well.

**Comment by ChengduLittleA** [comment]
Unfortunately I just failed to make the icon show on the tool bar. Here's what I did:

- Edited `symbolic-icons-extract.py`
- Set the group name to `mypaint-lock-radius`
- Add `mypaint-lock-radius-symbolic` to my action.
- Add the action to the tool bar python.
- it doesn't show up, I called `/svg/symbolic-icons-extract.py` but doesn't seem to produce that icon :(

Need some help I think...

**Comment by odysseywestra** [comment]
If you can help here, could you provide me a diff of what you changed with `symbolic-icons-extract.py`. I'm assuming you fixed the error with the getchildren() which was depreciated in python 3.2.

If I recall, the script needs to be run before you compile. Also a you need a packaged called `scour` installed. The py script use the ID and Label of the object properties of the group so everything can be on the icons layer without having to nest them under. The icon and the grey out border background just needs to be grouped together and you need to edit the object properties under ID with the icon name for the 16px and then append `:24` for the 24px. The script will add the `-symbolic` and move it to the appropriate folder if I recall. 

If the script is not working, you can just manually do it yourself. The script is mainly there to automate it if a large number of them are changed. Just make sure the icons are 16x16 and 24x24 and the `-symbolic` is added to the icon name. The 16px symbolic icons are under `/desktop/icons/hicolor/scalable/actions`, and the 24px icons are under `/desktop/icons/hicolor/24x24/actions`.

**Comment by ChengduLittleA** [comment]
Funny enough I didn't change `symbolic-icons-extract.py`, it doesn't seem to contain any list that needs to be modified. I'm not sure about the `getchildren()` call, but I didn't really see any errors when running the script.

With `scour` installed it still doesn't seem to generate that icon.

I'll try manually later today. Thanks for the info :D
